teps/0121-unified-retries-for-taskruns-and-customruns.md

```diff
@@ -0,0 +1,152 @@
+
+An [alternative](#alternatives) is removing support for retries from TaskRun and CustomRuns, make it a concern of PipelineRun.
+
+Another [alternative](#alternatives) is making retries a specialized field in each CustomRun individually instead of a common field for all CustomRuns.
+
+Another caveat is updating `retriesStatus` can lead to Large CRD issue, which can be alleviated by reducing the information we stored in `retriesStatus`, like removing taskSpec etc. We can design on it (similar to [TEP-0100](https://github.com/tektoncd/community/blob/main/teps/0100-embedded-taskruns-and-runs-status-in-pipelineruns.md)) when it becomes an actual issue.
+
+## Design Details
+
+### API Change
+#### New `retries` field in TaskRun.Spec
+
+```golang
+// TaskRunSpec defines the desired state of TaskRun
+type TaskRunSpec struct {
+  // ...
+  Retries int `json:"retries,omitempty"`
+  // ...
+}
+```
+
+Implements Retries in TaskRun reconciler, use `retriesStatus` to decide whether the retries has been exhausted.
+
+We pass [PipelineTask.Retries](https://github.com/tektoncd/pipeline/blob/481def234d1e8715ea5f0299713592f311340286/pkg/apis/pipeline/v1beta1/pipeline_types.go#L187) to TaskRunSpec.Retries and CustomRunSpec.Retries to retry TaskRuns and CustomRuns belonging to a parent PipelineRun.
+
+## Alternatives
+
+### Support retries for TaskRun and CustomRun in PipelineRun Controller
+
+Remove support for retries in TaskRun and CustomRun, make retries a PipelineRun concern; move logic for retries to PipelineRun reconciler and create new Runs at each attempt (remove the RetriesStatus); Pipeline Task Timeout applies to each retry individually.
+
+In this way, we don't need to worry about whether the retries is implemented correctly in CustomRun.
+
+However, TaskRun and CustomRun **can't retry on their own**, as Runtime Objects that can run on their own, we expect they can also be retried to avoid transient errors. There are use cases implying that we need it:
+  - Run: [PipelineLoop](https://github.com/tektoncd/experimental/blob/main/pipeline-loops/examples/pipelinespec-with-run-arrary-value.yaml)
+  - TaskRuns
+    - https://github.com/tektoncd/catalog/tree/main/task/send-to-webhook-slack/0.1
+    - https://github.com/tektoncd/catalog/tree/main/task/sendmail/0.1
+
+In the meantime users would be confused about which retries field to use when the CustomRun has both `specialized-retries` field as well as the commonly used `retries` field.
+
+Moreover, it does not allow custom run controllers to implement specialized retries (as described in the first alternative in [TEP-0069](https://github.com/tektoncd/community/blob/main/teps/0069-support-retries-for-custom-task-in-a-pipeline.md#alternatives))
```

Comment:
I think this doesn't prevent custom run controllers from implementing specialized retries, the problem is just that users wouldn't know which retries option to use.

Reply:
That's true, thanks for the comment! removed this line.